src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+            cur_pair[1][a] = -1;
+
+        return op;
+    }
+
+    instruction_ref finalize_output(const onnx_parser::node_info& info,
+                                    instruction_ref op,
+                                    const int_mat& map_mat,
+                                    int_mat& cur_pair) const
+    {
+        if(any_of(map_mat.back(), [](auto i) { return i >= 0; }))
+        {
+            cur_pair[1] = map_mat.back();
+            int_vec red;
+            for(int d = 0; d < map_mat[0].size(); ++d)
+            {
+                if(cur_pair[0][d] > 0 and cur_pair[1][d] == -1)
+                    red.push_back(d);
+            }
+
+            op = apply_reduce_sum_op(info, op, red, cur_pair[1]);
+        }
+
+        return transpose_squeeze(info, cur_pair, op, map_mat.back());
+    }
+
+    instruction_ref unsqueeze_transpose(const onnx_parser::node_info& info,
+                                        int_mat& cur_pair,
+                                        instruction_ref op) const
+    {
+        int_vec unsq_axes;
+        std::vector<std::tuple<int, int>> perm;
+
+        for(auto i = 0; i < cur_pair[1].size(); ++i)
+        {
+            if(cur_pair[1][i] == -1)
+                unsq_axes.push_back(i);
+            else
+                perm.push_back({cur_pair[1][i], i});
+        }
+        op = info.add_instruction(make_op("unsqueeze", {{"axes", unsq_axes}}), op);
+
+        std::sort(perm.begin(), perm.end(), [](auto lhs, auto rhs) {
+            return std::get<0>(lhs) < std::get<0>(rhs);
+        });
+
+        int_vec new_perm(cur_pair[1].size());
+        std::iota(new_perm.begin(), new_perm.end(), 0);
+
+        for(auto i = 0, p = 0; i < cur_pair[1].size(); ++i)
+        {
+            if(cur_pair[1][i] == -1)
+                continue;
+
+            new_perm[std::get<1>(perm[p++])] = i;
+        }
+
+        op = apply_transpose_op(info, op, new_perm, cur_pair[1]);
+
+        return op;
+    }
+
+    instruction_ref transpose_squeeze(const onnx_parser::node_info& info,
+                                      int_mat& cur_pair,
+                                      instruction_ref op,
+                                      int_vec row_output) const
+    {
+        std::vector<std::tuple<int, int>> perm;
+        int_vec sq;
+
+        for(auto i = 0; i < row_output.size(); ++i)
+        {
+            if(row_output[i] == -1)
+                sq.push_back(i);
+            else
+                perm.push_back({row_output[i], i});
+        }
+
+        std::sort(perm.begin(), perm.end(), [](auto lhs, auto rhs) {
+            return std::get<0>(lhs) < std::get<0>(rhs);
+        });
+
+        int_vec new_perm(cur_pair[1].size());
+        std::iota(new_perm.begin(), new_perm.end(), 0);
+
+        for(auto i = 0, p = 0; i < row_output.size(); ++i)
+        {
+            if(row_output[i] == -1)
+                continue;
+
+            new_perm[i] = std::get<1>(perm[p++]);
+        }
```

Comment:
Use `transform_if` for this.

Reply:
The problem is that `transform_if` doesn't advance the output iterator unless the predicate is satisfied. Besides that, a separate counter is used to index `perm`.

Reply:
I think `sort_permutation` can be used here as well.